File: bestia-zoneserver/src/main/java/net/bestia/zoneserver/actor/zone/SendClientsInRangeActor.java
package net.bestia.zoneserver.actor.zone;

import java.util.List;
import java.util.Objects;
import java.util.Optional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import net.bestia.entity.EntityService;
import net.bestia.entity.component.PositionComponent;
import net.bestia.messages.EntityJsonMessage;
import net.bestia.messages.JsonMessage;
import net.bestia.model.geometry.Point;
import net.bestia.model.geometry.Rect;
import net.bestia.zoneserver.actor.SpringExtension;
import net.bestia.zoneserver.map.MapService;
import net.bestia.zoneserver.service.PlayerEntityService;

/***
 * If a {@link EntityJsonMessage} is received by this actor it will check if the
 * given entity contains a {@link PositionComponent} and if it does it will
 * detect all active player entities in the update range of the game and forward
 * the message to them via a {@link SendClientActor}.
 * 
 * @author Thomas Felix
 *
 */
@Component
@Scope("prototype")
public class SendClientsInRangeActor extends AbstractActor {

	private final LoggingAdapter LOG = Logging.getLogger(getContext().system(), this);
	public static final String NAME = "sendToClientsInRange";

	private final PlayerEntityService playerEntityService;
	private final EntityService entityService;
	private final ActorRef sendClient;

	@Autowired
	public SendClientsInRangeActor(PlayerEntityService playerEntityService,
			EntityService entityService) {

		this.entityService = Objects.requireNonNull(entityService);
		this.playerEntityService = Objects.requireNonNull(playerEntityService);
		this.sendClient = SpringExtension.actorOf(getContext(), SendClientActor.class);
		
	}

	@Override
	public Receive createReceive() {
		return receiveBuilder()
				.match(EntityJsonMessage.class, this::sendToActiveInRange)
				.build();
	}

	/**
	 * Sends to all active players in range. Maybe automatically detecting this
	 * method by message type does not work out. Then we might need to create a
	 * whole new actor only responsible for sending active range messages. Might
	 * be better idea anyways.
	 * 
	 * @param msg
	 */
	private void sendToActiveInRange(EntityJsonMessage msg) {

		// Get position of the entity.
		final Optional<PositionComponent> posComp = entityService.getComponent(msg.getEntityId(),
				PositionComponent.class);

		if (!posComp.isPresent()) {
			LOG.warning("Position component of entity in message {} not present. Can not send range update.", msg);
			return;
		}

		final Point pos = posComp.get().getPosition();
		final Rect updateRect = MapService.getUpdateRect(pos);

		final List<Long> activeIds = playerEntityService.getActiveAccountIdsInRange(updateRect);

		for (Long activeId : activeIds) {		
			final JsonMessage newMsg = msg.createNewInstance(activeId);		
			sendClient.tell(newMsg, getSelf());
		}
	}
}

File: bestia-zoneserver/src/main/kotlin/net/bestia/zoneserver/actor/client/SendClientsInRangeActor.kt
package net.bestia.zoneserver.actor.client

import akka.actor.ActorRef
import mu.KotlinLogging
import net.bestia.messages.AccountMessage
import net.bestia.messages.client.ClientEnvelope
import net.bestia.zoneserver.actor.routing.MessageApi
import net.bestia.zoneserver.actor.Actor
import net.bestia.zoneserver.actor.BQualifier
import net.bestia.zoneserver.entity.component.PositionComponent
import net.bestia.zoneserver.actor.entity.awaitEntityResponse
import net.bestia.zoneserver.actor.routing.DynamicMessageRoutingActor
import net.bestia.zoneserver.entity.EntityCollisionService
import net.bestia.zoneserver.entity.Entity
import net.bestia.zoneserver.entity.component.OwnerComponent
import net.bestia.zoneserver.map.MapService
import org.springframework.beans.factory.annotation.Qualifier

private val LOG = KotlinLogging.logger { }

internal data class SendInRange(
    val entity: Entity,
    val content: AccountMessage
)

/***
 * If a [EntityJsonMessage] is received by this actor it will check if the
 * given entity contains a [PositionComponent] and if it does it will
 * detect all active player entities in the update range of the game and forward
 * the message to them via a [SendToClientActor].
 *
 * @author Thomas Felix
 */
@Actor
class SendClientsInRangeActor(
    private val entityCollisionService: EntityCollisionService,
    private val messageApi: MessageApi,
    @Qualifier(BQualifier.CLIENT_FORWARDER)
    private val sendClientActor: ActorRef
) : DynamicMessageRoutingActor() {

  override fun createReceive(builder: BuilderFacade) {
    builder.matchRedirect(SendInRange::class.java, this::handleSendToActiveInRange)
  }

  /**
   * Sends to all active players in range. Maybe automatically detecting this
   * method by message type does not work out. Then we might need to create a
   * whole new actor only responsible for sending active range messages. Might
   * be better idea anyways.
   */
  private fun handleSendToActiveInRange(msg: SendInRange) {
    LOG.trace { "Received: $msg" }

    val posComp = msg.entity.tryGetComponent(PositionComponent::class.java)
        ?: run {
          LOG.warn { "Position component of entity in message $msg not present. Can not send range update." }
          return
        }

    val updateRect = MapService.getUpdateRect(posComp.position)
    val activeIds = entityCollisionService.getAllCollidingEntityIds(updateRect)

    awaitEntityResponse(messageApi, context, activeIds) { entities ->
      val playerAccountIds = entities.all
          .mapNotNull { it.tryGetComponent(OwnerComponent::class.java)?.ownerAccountIds }
          .flatten()
          .toSet()
      playerAccountIds.forEach {
        sendClientActor.tell(ClientEnvelope(it, msg.content), self)
      }
    }
  }

  companion object {
    const val NAME = "sendToClientsInRange"
  }
}

